src/object.d

```diff
@@ -4881,3 +4881,90 @@ void __ArrayDtor(T)(T[] a)
     foreach_reverse (ref T e; a)
         e.__xdtor();
 }
+
+/***********************************************************
+ * The compiler lowers expressions of `cast(TTo[])TFrom[]` to
+ * this implementation.
+ * Params:
+ *  from = the array to reinterpret-cast
+ * Returns:
+ *  `from` reinterpreted as `TTo[]`
+ */
+TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @trusted
+{
+    const nbytes = from.length * TFrom.sizeof;
+
+    struct Array
+    {
+        size_t length;
+        void* ptr;
+    }
+
+    if ((nbytes % TTo.sizeof) != 0)
+    {
+        import core.internal.string : unsignedToTempString;
+        import core.stdc.stdlib : alloca;
+
+        const(char)[][8] msgComponents =
+        [
+            "Cannot cast `"
+            , TFrom.stringof
+            , "` to `"
+            , TTo.stringof
+            , "`, an array of size "
+            , unsignedToTempString(nbytes)
+            , " does not align on an array of size "
+            , unsignedToTempString((nbytes / TTo.sizeof) + (nbytes % TTo.sizeof))
+        ];
+
+        // convert discontiguous `msgComponents` to contiguous string on the stack
+        char[] msg;
+
+        auto array = cast(Array*)&msg;
+
+        foreach(m ; msgComponents)
+            array.length += m.length;
+        array.ptr = alloca(array.length);
+
+        size_t index = 0;
+        foreach(m ; msgComponents)
+            foreach(c; m)
+                msg[index++] = c;
+
+        version (D_Exceptions)
+        {
+            import core.exception : throwStaticError;
+            throwStaticError(cast(string)msg);
+        }
+        else
```

Comment:
It looks like D is moving in C++ land here:

> Divergent error handling has fractured the C++ community into incompatible dialects:
(1) C++ projects often ban even turning on compiler support for exception handling, but this means they are
not using Standard C++. Exceptions are required to use central C++ standard language features (e.g., con-
structors) and the C++ standard library. Yet in [SC++F 2018], __over half of C++ developers__ report that __excep-
tions are banned__ in part (32%) or all (20%) of their code, which means they are using a__ divergent language
dialect with different idioms__ (e.g., two-phase construction) and either a nonconforming standard library
dialect or none at all. We must make it possible for all C++ projects to at least turn on exception handling
support and use the standard language and library.

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r0.pdf

Reply:
Indeed.  That's what happens when a language is not designed from the beginning with freestanding, resource-constrained systems and scalability in mind.  The emergence of things like `@nogc`, betterC, etc. shows that D is backtracking, trying to undo some mistakes of the past.  GC should be an opt-in library feature and exceptions should not be rooted in `Object` or Java-style `class`-like types.  But that's what we have.  If we want D to scale to systems other than resource-abundant PCs and servers running full-featured operating systems, then we have to offer alternatives.  I hope future language decisions will keep principles of minimalism in mind.  Given [what's possible in D](https://forum.dlang.org/post/qjowkuwstewnmdunepiq@forum.dlang.org), I'd prefer if classes were just deprecated, and structs were give a few more features.